File: backend/internal/handlers/tag.go
package handlers

import (
	"github.com/gin-gonic/gin"
	"net/http"
	"online_library/backend/internal/dto"
	"online_library/backend/internal/models"
	"online_library/backend/internal/service"
	"strconv"
	"strings"
)

type TagHandler struct {
	tagService service.TagService
}

func NewTagHandler(tagService service.TagService) *TagHandler {
	return &TagHandler{tagService: tagService}
}

func (h *TagHandler) SearchTags(c *gin.Context) {
	query := strings.TrimSpace(c.Query("query"))

	tags, err := h.tagService.GetAllTags()
	if err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": "failed to fetch tags"})
		return
	}

	var filtered []models.Tag
	if query != "" {
		queryLower := strings.ToLower(query)
		for _, tag := range tags {
			if strings.Contains(strings.ToLower(tag.Name), queryLower) {
				filtered = append(filtered, tag)
			}
		}
	} else {
		filtered = tags
	}

	c.JSON(http.StatusOK, dto.ConvertTags(filtered))
}

func (h *TagHandler) GetTagByID(c *gin.Context) {
	id, _ := strconv.Atoi(c.Param("id"))

	tag, err := h.tagService.GetTagByID(id)
	if err != nil {
		c.JSON(http.StatusNotFound, gin.H{"error": "tag not found"})
		return
	}

	c.JSON(http.StatusOK, dto.TagDTO{
		ID:    tag.ID,
		Name:  tag.Name,
		Color: tag.Color,
	})
}

func (h *TagHandler) CreateTag(c *gin.Context) {
	var input dto.TagDTO
	if err := c.ShouldBindJSON(&input); err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"error": "invalid input"})
		return
	}

	tag := input.ToModel()
	if err := h.tagService.CreateTag(tag); err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": "failed to create tag"})
		return
	}

	input.ID = tag.ID
	c.JSON(http.StatusCreated, input)
}

func (h *TagHandler) UpdateTag(c *gin.Context) {
	id, _ := strconv.Atoi(c.Param("id"))

	var input dto.TagDTO
	if err := c.ShouldBindJSON(&input); err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"error": "invalid input"})
		return
	}

	input.ID = id
	tag := input.ToModel()

	if err := h.tagService.UpdateTag(tag); err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": "failed to update tag"})
		return
	}

	c.JSON(http.StatusOK, input)
}

func (h *TagHandler) DeleteTag(c *gin.Context) {
	id, _ := strconv.Atoi(c.Param("id"))
	if err := h.tagService.DeleteTag(id); err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": "failed to delete tag"})
		return
	}
	c.Status(http.StatusNoContent)
}

func (h *TagHandler) GetTagsByBookID(c *gin.Context) {
	bookID, _ := strconv.Atoi(c.Param("bookID"))

	tags, err := h.tagService.GetTagsByBookID(bookID)
	if err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": "failed to fetch tags"})
		return
	}

	c.JSON(http.StatusOK, dto.ConvertTags(tags))
}

func (h *TagHandler) AssignTagToBook(c *gin.Context) {
	var bt models.BookTag
	if err := c.ShouldBindJSON(&bt); err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"error": "invalid input"})
		return
	}
	if err := h.tagService.AssignTagToBook(&bt); err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": "failed to assign tag"})
		return
	}
	c.Status(http.StatusOK)
}

func (h *TagHandler) RemoveTagFromBook(c *gin.Context) {
	bookID, _ := strconv.Atoi(c.Query("book_id"))
	tagID, _ := strconv.Atoi(c.Query("tag_id"))
	if err := h.tagService.RemoveTagFromBook(bookID, tagID); err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": "failed to remove tag"})
		return
	}
	c.Status(http.StatusOK)
}

File: frontend/src/api/tag.js
import axios from '@/api/axios'

/**
 * Получает информацию о теге по его ID.
 *
 * @async
 * @param {number|string} id - Уникальный идентификатор тега
 * @returns {Promise<import('axios').AxiosResponse>} Ответ с данными тега
 *
 * @example
 * const response = await fetchTag(3);
 * console.log(response.data);
 */
export async function fetchTag(id) {
    return axios.get(`/tags/${id}`)
}

/**
 * Получает список книг, связанных с определённым тегом.
 *
 * @async
 * @param {number|string} id - Уникальный идентификатор тега
 * @returns {Promise<import('axios').AxiosResponse>} Ответ с книгами по тегу
 *
 * @example
 * const response = await fetchTagBooks(5);
 * console.log(response.data);
 */
export async function fetchTagBooks(id) {
    return axios.get(`/books/tag/${id}`)
}

/**
 * Выполняет поиск тегов по строке запроса.
 *
 * @async
 * @param {string} query - Поисковый запрос
 * @returns {Promise<import('axios').AxiosResponse>} Ответ со списком тегов
 *
 * @example
 * const response = await searchTags('history');
 */
export async function searchTags(query) {
    return axios.get(`/tags?query=${encodeURIComponent(query)}`)
}

/**
 * Создаёт новый тег.
 *
 * @async
 * @param {{ name: string }} data - Данные нового тега
 * @returns {Promise<import('axios').AxiosResponse>} Ответ с созданным тегом
 */
export async function createTag(data) {
    return axios.post('/tags', data)
}

/**
 * Обновляет существующий тег.
 *
 * Требуются права администратора.
 *
 * @async
 * @param {number|string} id - ID тега
 * @param {{ name: string }} data - Обновлённые данные
 * @returns {Promise<import('axios').AxiosResponse>} Ответ с обновлённым тегом
 */
export async function updateTag(id, data) {
    return axios.put(`/tags/${id}`, data)
}

/**
 * Удаляет тег.
 *
 * Требуются права администратора.
 *
 * @async
 * @param {number|string} id - ID тега
 * @returns {Promise<import('axios').AxiosResponse>} Ответ сервера
 */
export async function deleteTag(id) {
    return axios.post(`/tags/${id}/delete`)
}

/**
 * Получает все теги, назначенные книге.
 *
 * @async
 * @param {number|string} bookId - ID книги
 * @returns {Promise<import('axios').AxiosResponse>} Ответ со списком тегов
 */
export async function getTagsByBookId(bookId) {
    return axios.get(`/tags/book/${bookId}`)
}

/**
 * Назначает тег книге.
 *
 * Требуются права владельца книги или администратора.
 *
 * @async
 * @param {number} tagID - ID тега
 * @param {number} bookID - ID книги
 * @returns {Promise<import('axios').AxiosResponse>} Ответ сервера
 */
export async function assignTagToBook(tagID, bookID) {
    return axios.post('/tags/assign', {
        tag_id: tagID,
        book_id: bookID
    })
}

/**
 * Удаляет тег из книги.
 *
 * Требуются права владельца книги или администратора.
 *
 * @async
 * @param {number} tagID - ID тега
 * @param {number} bookID - ID книги
 * @returns {Promise<import('axios').AxiosResponse>} Ответ сервера
 */
export async function removeTagFromBook(tagID, bookID) {
    return axios.post('/tags/remove', {
        tag_id: tagID,
        book_id: bookID
    })
}